provider/application.go

```diff
@@ -0,0 +1,191 @@
+			},
+		},
+	}
+}
+
+func (a *application) setResourceData(d *schema.ResourceData) error {
+	d.SetId(a.ID)
+	err := d.Set("name", a.Name)
+	if err != nil {
+		return err
+	}
+	err = d.Set("email", a.Email)
+	if err != nil {
+		return err
+	}
+	err = d.Set("repo_type", a.RepoType)
+	if err != nil {
+		return err
+	}
+	err = d.Set("repo_project_key", a.RepoProjectKey)
+	if err != nil {
+		return err
+	}
+	err = d.Set("repo_slug", a.RepoSlug)
+	if err != nil {
+		return err
+	}
+	err = d.Set("instance_port", a.InstancePort)
+	if err != nil {
+		return err
+	}
+	err = d.Set("platform_health_only", a.PlatformHealthOnly)
+	if err != nil {
+		return err
+	}
+	err = d.Set("platform_health_only_show_override", a.PlatformHealthOnlyShowOverride)
+	if err != nil {
+		return err
+	}
+	err = d.Set("enable_restart_running_executions", a.EnableRestartRunningExecutions)
+	if err != nil {
+		return err
+	}
+	return d.Set("cloud_providers", a.CloudProviders)
+}
+
+// ApplicationFromResourceData get application from resource data
+func applicationFromResourceData(application *client.Application, d *schema.ResourceData) {
+	application.Name = d.Get("name").(string)
+	application.Email = d.Get("email").(string)
+	application.RepoType = d.Get("repo_type").(string)
+	application.RepoProjectKey = d.Get("repo_project_key").(string)
+	application.RepoSlug = d.Get("repo_slug").(string)
+	application.CloudProviders = applicationCloudProvidersFromResourceData(d)
+	application.ProviderSettings = applicationProviderSettingsFromResourceData(d)
+	application.InstancePort = d.Get("instance_port").(int)
+	application.PlatformHealthOnly = d.Get("platform_health_only").(bool)
+	application.PlatformHealthOnlyShowOverride = d.Get("platform_health_only_show_override").(bool)
+	application.EnableRestartRunningExecutions = d.Get("enable_restart_running_executions").(bool)
+}
+
+func applicationCloudProvidersFromResourceData(d *schema.ResourceData) string {
+	cloudProvidersInterface, ok := d.GetOk("cloud_providers")
+	if !ok {
+		return ""
+	}
+
+	cloudProviders := []string{}
+	for _, cloudProvider := range cloudProvidersInterface.([]interface{}) {
+		cloudProviders = append(cloudProviders, cloudProvider.(string))
+	}
+	return strings.Join(cloudProviders, ",")
+}
+
+func applicationProviderSettingsFromResourceData(d *schema.ResourceData) *client.ProviderSettings {
+	providerSettingsInterface, ok := d.GetOk("provider_settings")
+	if !ok {
+		return nil
+	}
+
+	clientProviderSettings := &client.ProviderSettings{}
+	for _, providerSettingInterface := range providerSettingsInterface.([]interface{}) {
+		providerSetting := providerSettingInterface.(map[string]interface{})
+		clientProviderSettings.AWS = awsProviderSettingsFromResourceData(providerSetting)
+	}
+	return clientProviderSettings
+}
+
+func awsProviderSettingsFromResourceData(providerSetting map[string]interface{}) *client.AwsProviderSettings {
+	for _, awsInterface := range providerSetting["aws"].([]interface{}) {
```

Comment:
same here -- is it possible we could typeswitch on this or marshal it into a more defined structure?